File: 55-jump-game/jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        /*
        2,3,1,1,4
        2 4

        3,2,1,0,4
        3 3 3 3
        Is current index <= a position we already found?
        if no, skip

        3,2,1,0,1,2
        3 3 3 3 

        */
        if (nums.length == 1) return true;

        int maxFound = -1;

        for (int i = 0; i < nums.length - 1; i++) {
            if (i == 0 || i <= maxFound) {
                int jumpPos = i + nums[i];
                // Cuz that's the maximum jump pos,
                // so if you can jump way more than the last element, then you can def jump to the last element
                if (jumpPos >= nums.length - 1) {
                    return true;
                }
                maxFound = Math.max(maxFound, jumpPos);
            }
        }
        return false;
    }
}

File: 55-jump-game/jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        """
        [2,3,1,1,4]

        2 -> most you can do is index 2 (3, 1)

        3 -> most you can do is index 4 (1, 1, 4)
        you're at the end

        [3, 2, 1, 0, 4]
        3 -> index 3
        2 -> index 3
        1 -> index 3
        0 -> index 3
        4 [at index 4] -> can't reach because max jump is index 3

        """
        max_jump = 0
        for i in range(len(nums)):
            if i > max_jump:
                return False

            max_jump = max(max_jump, i + nums[i])
            
        return True